configserver-client/src/main/java/ai/vespa/hosted/client/ConfigServerClient.java

```diff
@@ -0,0 +1,148 @@
+
+/**
+ * @author jonmv
+ */
+public interface ConfigServerClient extends AutoCloseable {
+
+    /** Creates a builder for sending the given method, using the specified host strategy. */
+    RequestBuilder send(HostStrategy hosts, Method method);
+
+    /** Builder for a request against a given set of hosts, using this config server client. */
+    interface RequestBuilder {
+
+        /** Sets the request path. */
+        RequestBuilder at(String... pathSegments);
+
+        /** Sets the request body as UTF-8 application/json. */
+        RequestBuilder body(byte[] json);
+
+        /** Sets the request body. */
+        RequestBuilder body(HttpEntity entity);
+
+        /** Sets the parameter key/values for the request. Number of arguments must be even. */
+        RequestBuilder parameters(String... pairs);
+
+        /** Overrides the default socket read timeout of the request. {@code Duration.ZERO} gives infinite timeout. */
+        RequestBuilder timeout(Duration timeout);
+
+        /** Overrides the default socket read timeout of the request. {@code null} allows infinite timeout. */
+        RequestBuilder config(RequestConfig config);
+
+        /**
+         * Sets custom retry/failure logic for this.
+         * <p>
+         * Exactly one of the arguments (response, exception) are non-null.
+         * Return a value to have that returned to the caller;
+         * throw a {@link RetryException} to have the request retried; or
+         * throw any other unchecked exception to have this propagate out to the caller.
+         * The caller must close the provided response, if any.
+         */
+        <T> T handle(BiFunction<ClassicHttpResponse, IOException, T> handler) throws UncheckedIOException;
+
+        /** Sets the response body mapper for this, for successful requests. */
+        <T> T read(Function<byte[], T> mapper) throws UncheckedIOException, ConfigServerException;
+
+        /** Discards the response, but throws if the response is unsuccessful. */
+        void discard() throws UncheckedIOException, ConfigServerException;
+
+        /** Returns the raw input stream of the response, if successful. The caller must close the returned stream. */
+        InputStream stream() throws UncheckedIOException, ConfigServerException;
+
+    }
+
+    /** Exception wrapper that signals retries should be attempted. */
+    final class RetryException extends RuntimeException {
+
+        public RetryException(IOException cause) {
+            super(requireNonNull(cause));
+        }
+
+        public RetryException(RuntimeException cause) {
+            super(requireNonNull(cause));
+        }
+
+    }
+
+    /** What host(s) to try for a request, in what order. A host may be specified multiple times, for retries.  */
+    @FunctionalInterface
+    interface HostStrategy extends Iterable<URI> {
+
+        /** Attempts each request once against each listed host. */
+        static HostStrategy ordered(List<URI> hosts) {
```

Comment:
Is it only the hostname that is different among the supplied URIs? Consider taking a list of hostnames and the other fixed URI parts as additional parameters

Reply:
Hmm ... I guess that's true right now. Will consider it.